server/embed/config_test.go

```diff
@@ -89,6 +91,117 @@ func TestConfigFileOtherFields(t *testing.T) {
 	assert.Equal(t, false, cfg.SocketOpts.ReuseAddress, "ReuseAddress does not match")
 }
 
+func TestConfigFileFeatureGates(t *testing.T) {
+	testCases := []struct {
+		name                                   string
+		featureGatesJSON                       string
+		experimentalStopGRPCServiceOnDefrag    bool
+		setExperimentalStopGRPCServiceOnDefrag bool
+		expectErr                              bool
+		expectedFeatures                       map[featuregate.Feature]bool
+	}{
+		{
+			name: "default",
+			expectedFeatures: map[featuregate.Feature]bool{
+				features.DistributedTracing:      false,
+				features.StopGRPCServiceOnDefrag: false,
+			},
+		},
+		{
+			name:                                   "cannot set both experimental flag and feature gate flag",
+			featureGatesJSON:                       "StopGRPCServiceOnDefrag=true",
+			experimentalStopGRPCServiceOnDefrag:    false,
+			setExperimentalStopGRPCServiceOnDefrag: true,
+			expectErr:                              true,
+		},
+		{
+			name:                                   "ok to set different experimental flag and feature gate flag",
+			featureGatesJSON:                       "DistributedTracing=false",
+			experimentalStopGRPCServiceOnDefrag:    true,
+			setExperimentalStopGRPCServiceOnDefrag: true,
+			expectedFeatures: map[featuregate.Feature]bool{
+				features.DistributedTracing:      false,
+				features.StopGRPCServiceOnDefrag: true,
+			},
+		},
+		{
+			name:                                   "can set feature gate to true from experimental flag",
+			experimentalStopGRPCServiceOnDefrag:    true,
+			setExperimentalStopGRPCServiceOnDefrag: true,
+			expectedFeatures: map[featuregate.Feature]bool{
+				features.StopGRPCServiceOnDefrag: true,
+			},
+		},
+		{
+			name:                                   "can set feature gate to false from experimental flag",
+			experimentalStopGRPCServiceOnDefrag:    false,
+			setExperimentalStopGRPCServiceOnDefrag: true,
+			expectedFeatures: map[featuregate.Feature]bool{
+				features.StopGRPCServiceOnDefrag: false,
+			},
+		},
+		{
+			name:             "can set feature gate to true from feature gate flag",
+			featureGatesJSON: "StopGRPCServiceOnDefrag=true",
+			expectedFeatures: map[featuregate.Feature]bool{
+				features.StopGRPCServiceOnDefrag: true,
+			},
+		},
+		{
+			name:             "can set feature gate to false from feature gate flag",
+			featureGatesJSON: "StopGRPCServiceOnDefrag=false",
+			expectedFeatures: map[featuregate.Feature]bool{
+				features.StopGRPCServiceOnDefrag: false,
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var b []byte
+			var err error
+			if tc.setExperimentalStopGRPCServiceOnDefrag {
```

Comment:
Don't understand why `setExperimentalStopGRPCServiceOnDefrag` exist, or what it does.

Reply:
Removed. I cannot use the `map[string]string` because yaml will marshal the `bool` as `string` which would cause parsing error later, so have to use the 2 struct approach.